### PR TITLE
Add connection quality (latency and reliability) based routing

### PR DESCRIPTION
**Pull request by jmacAJ**
- added multiple endpoints capability per model
- fastest endpoint will be chosen at request time
- endpoints can be marked unhealthy if they have errors
- all endpoint status ages out - so will turn healthy again after time and successful requests
- simplified some of the parameter hell
- renamed pathwayPrompter to modelExecutor
- renamed request to requestExecutor
- upgraded the requestMonitor functionality significantly
- reduced parameters required for calls throughout

### CODE REVIEW

File: config.js [hunk 1]
```diff
@@ -264,17 +264,41 @@ const buildPathways = async (config) => {
 
 // Build and load models to config
 const buildModels = (config) => {
-    const { models } = config.getProperties();
+    let { models } = config.getProperties();
```
Comment (ademgcl) [inline config.js:267]: we don't seem to change var, so const?

File: lib/requestExecutor.js [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    if (!model || !Array.isArray(model.endpoints) || model.endpoints.length === 0) {
+        return null;
+    } else {
+        logger.debug(`Selecting endpoint for model ${model.name}...`);
+        if (model.endpoints.length === 1) {
+            logger.debug(`Only one endpoint for model ${model.name}. No selection required.`);
+            return model.endpoints[0];
+        }
+
+        let healthyEndpoints = model.endpoints.filter(endpoint => endpoint.monitor.healthy);
```
Comment (ademgcl) [inline lib/requestExecutor.js:96]: i think this is great for the first version. i'd still vote for some kind of pq approach then we could possibly get rid of several ifs and pick in o1, but just a thought

File: server/graphql.js [hunk 3]
```diff
@@ -97,7 +97,7 @@ const getResolvers = (config, pathways) => {
             // add shared state to contextValue
             contextValue.pathway = pathway;
             contextValue.config = config;
-            return pathway.rootResolver(parent, args, contextValue, info);
+                        return pathway.rootResolver(parent, args, contextValue, info);
```
Comment (ademgcl) [inline server/graphql.js:100]: seemingly unintentional space?

**Review by ademgcl — APPROVED** [review]
LGTM

File: server/modelExecutor.js [hunk 4]
```diff
@@ -16,59 +18,59 @@ import OpenAIImagePlugin from './plugins/openAiImagePlugin.js';
...
-    constructor(config, pathway, modelName, model) {
+class ModelExecutor {
+    constructor(pathway, model) {
 
         let plugin;
 
         switch (model.type) {
             case 'OPENAI-CHAT':
-                plugin = new OpenAIChatPlugin(config, pathway, modelName, model);
+                plugin = new OpenAIChatPlugin(pathway, model);
```
Comment (hammadsaj) [inline server/modelExecutor.js:28]: Always nice to have to pass fewer parameters

File: lib/requestDurationEstimator.js
Comment (hammadsaj) [inline lib/requestDurationEstimator.js:1]: 👍

**Review by hammadsaj — APPROVED** [review]

**Pull request merged**